src/pgen/cluster/jet_coords.hpp

```diff
@@ -0,0 +1,115 @@
+                     const parthenon::Real phi_jet_axis)
+      : cos_theta_jet_axis_(cos(theta_jet_axis)),
+        sin_theta_jet_axis_(sin(theta_jet_axis)), cos_phi_jet_axis_(cos(phi_jet_axis)),
+        sin_phi_jet_axis_(sin(phi_jet_axis)) {}
+
+  // Convert simulation cartesian coordinates to jet cylindrical coordinates
+  KOKKOS_INLINE_FUNCTION void
+  SimCartToJetCylCoords(const parthenon::Real x_sim, const parthenon::Real y_sim,
+                        const parthenon::Real z_sim, parthenon::Real &r_jet,
+                        parthenon::Real &cos_theta_jet, parthenon::Real &sin_theta_jet,
+                        parthenon::Real &h_jet) const __attribute__((always_inline)) {
+
+    // Position in jet-cartesian coordinates
+    const parthenon::Real x_jet = x_sim * cos_phi_jet_axis_ * cos_theta_jet_axis_ +
+                                  y_sim * sin_phi_jet_axis_ -
+                                  z_sim * sin_theta_jet_axis_ * cos_phi_jet_axis_;
+    const parthenon::Real y_jet = -x_sim * sin_phi_jet_axis_ * cos_theta_jet_axis_ +
+                                  y_sim * cos_phi_jet_axis_ +
+                                  z_sim * sin_phi_jet_axis_ * sin_theta_jet_axis_;
+    const parthenon::Real z_jet =
+        x_sim * sin_theta_jet_axis_ + z_sim * cos_theta_jet_axis_;
+
+    // Position in jet-cylindrical coordinates
+    r_jet = sqrt(pow(fabs(x_jet), 2) + pow(fabs(y_jet), 2));
+    cos_theta_jet = x_jet / r_jet;
+    sin_theta_jet = y_jet / r_jet;
+    h_jet = z_jet;
+  }
+
+  // Convert jet cylindrical vector to simulation cartesian vector
+  KOKKOS_INLINE_FUNCTION void JetCylToSimCartVector(
+      const parthenon::Real cos_theta_jet, const parthenon::Real sin_theta_jet,
+      const parthenon::Real v_r_jet, const parthenon::Real v_theta_jet,
+      const parthenon::Real v_h_jet, parthenon::Real &v_x_sim, parthenon::Real &v_y_sim,
+      parthenon::Real &v_z_sim) const __attribute__((always_inline)) {
+    // The vector in jet-cartesian coordinates
+    const parthenon::Real v_x_jet = v_r_jet * cos_theta_jet - v_theta_jet * sin_theta_jet;
+    const parthenon::Real v_y_jet = v_r_jet * sin_theta_jet + v_theta_jet * cos_theta_jet;
+    const parthenon::Real v_z_jet = v_h_jet;
+
+    // Multiply v_jet by the DCM matrix to take Jet cartesian to Simulation Cartesian
+    v_x_sim = v_x_jet * cos_phi_jet_axis_ * cos_theta_jet_axis_ -
+              v_y_jet * sin_phi_jet_axis_ * cos_theta_jet_axis_ +
+              v_z_jet * sin_theta_jet_axis_;
+    v_y_sim = v_x_jet * sin_phi_jet_axis_ + v_y_jet * cos_phi_jet_axis_;
+    v_z_sim = -v_x_jet * sin_theta_jet_axis_ * cos_phi_jet_axis_ +
+              v_y_jet * sin_phi_jet_axis_ * sin_theta_jet_axis_ +
+              v_z_jet * cos_theta_jet_axis_;
+  }
+};
+/************************************************************
+ * Jet Coordinates Factory Class
+ * A factory for creating JetCoords objects given a time
+ * Keeps track of the precession
```

Comment:
I don't understand some of this question  -- `phi_dot` is constant for a precessing jet.

The way the `JetCoords` class is used is that we create a new one every time we need one. This allow us to execute some of the jet trigonometry functions once at creation rather than at every point. The current implementation is an optimization.

Reply:
Gotcha. Then I misunderstood the comment in the code.
I thought that "keeping track" meant that the object is persistent and updated (according to some prescribed manner so that there's a track record).
This is also why I asked about the varying `phi_dot` as for a constant `phi_dot` things can be calculated at once (as you also say).